docs/writing-tests/making-a-testing-plan.md

```diff
@@ -0,0 +1,496 @@
+shared by many different APIs.
+
+In general, you should start by writing only "surface tests" for the nested
+algorithms. That means only verifying that they exhibit the basic behavior you
+are expecting.
+
+It's definitely important to test exhaustively, but it's just as important to
+do so in a structured way. Reach out to the test suite's maintainers to learn
+if and how they have already tested those algorithms. In many cases, it's
+acceptable to test them in just one place (and maybe through a different API
+entirely), and rely only on surface-level testing everywhere else. While it's
+always possible for more tests to uncover new bugs, the chances may be slim.
+The time we spend writing tests is highly valuable, so we have to be efficient!
+
+*Example:* The following algorithm from [the DOM
+standard](https://dom.spec.whatwg.org/) powers
+[`document.querySelector`](https://developer.mozilla.org/en-US/docs/Web/API/Document/querySelector):
+
+> To **scope-match a selectors string** *selectors* against a *node*, run these
+> steps:
+>
+> 1. Let *s* be the result of [parse a
+>    selector](https://drafts.csswg.org/selectors-4/#parse-a-selector)
+>    *selectors*.
+> 2. If *s* is failure, then
+>    [throw](https://heycam.github.io/webidl/#dfn-throw) a
+>    "[`SyntaxError`](https://heycam.github.io/webidl/#syntaxerror)"
+>    [DOMException](https://heycam.github.io/webidl/#idl-DOMException).
+> 3. Return the result of [match a selector against a
+>    tree](https://drafts.csswg.org/selectors-4/#match-a-selector-against-a-tree)
+>    with *s* and *node*'s
+>    [root](https://dom.spec.whatwg.org/#concept-tree-root) using [scoping
+>    root](https://drafts.csswg.org/selectors-4/#scoping-root) *node*.
+
+As described earlier in this guide, we'd certainly want to test the branch
+regarding the parsing failure. However, there are many ways a string might fail
+to parse--should we verify them all in the tests for `document.querySelector`?
+What about `document.querySelectorAll`? Should we test them all there, too?
+
+The answers depend on the current state of the test suite: whether or not tests
+for selector parsing exist and where they are located. That's why it's best to
+confer with the people who are maintaining the tests.
+
+### Avoid excessive breadth
+
+When the set of input values is finite, it can be tempting to test them all
+exhaustively. When the set is very large, test authors can reduce repetition by
+defining tests programmatically in loops.
+
+Using advanced control flow techniques to dynamically generate tests can
+actually *reduce* test quality. It may obscure the intent of the tests since
+readers have to mentally "unwind" the iteration to determine what is actually
+being verified. The practice is more susceptible to bugs. These bugs may not be
+obvious--they may not cause failures, and they may exercise fewer cases than
+intended. Finally, tests authored using this approach often have execution time
```

Comment:
I agree with @zcorpan in https://github.com/web-platform-tests/wpt/pull/19628/commits/10ae79576bcbc447bedaf7bb458537d4ac109e96#r337517178:
> I think it's easier to understand the problem by being more direct: When generating tests, it's easy to end up with a *lot* of tests which introduces the problem that running the tests takes too much time.

Reply:
Trying this:

> Finally, tests authored using this approach often take a relatively long time to complete, and that puts a burden on people who collect test results in large numbers.